### PR TITLE
ENH: Add weights parameter to stats.gmean

### LINKED ISSUE #13065: Weighted geometric mean
I propose the addition of a `weights` argument to `scipy.stats.gmean`, to support the calculation of the weighted geometric mean according to https://en.wikipedia.org/wiki/Weighted_geometric_mean. If there is interest, then I could create a PR.

### PR DESCRIPTION
**Pull request by maxfrei750**
I propose the addition of a `weights` argument to `scipy.stats.gmean`, to support the calculation of the weighted geometric mean according to https://en.wikipedia.org/wiki/Weighted_geometric_mean.

#### What does this implement/fix?
Allows the calculation of the weighted geometric mean.

#### Additional information
Docstring inspired by https://github.com/scipy/scipy/pull/6931/files#diff-764f0cedd33e98726d12c935495b80333a11e877a5a270d8ce53c9db152dd976R382-R384

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks @maxfrei750. gh-6931 is related indeed, good to review discussion on weights there.

**Comment by rgommers** [comment]
See https://github.com/scipy/scipy/pull/6931#issuecomment-270817843 with multiple interpretations of weights. Could you expand on the interpretation used here @maxfrei750?

Note that this will also need a unit test.

**Comment by maxfrei750** [comment]
I don't think that the implementation necessarily implies a certain interpretation. IMHO it depends on the supplied weights. I think this is also supported by the linked Wikipedia article, since it does not explicitly mention an interpretation.

> Note that this will also need a unit test.

If this PR is generally welcome, then I'll of course be happy to create a unit test.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:418]:
```suggestion
        weights = np.array(weights, dtype=dtype, copy=False)
```

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:426]: Please raise an appropriate error type.

Comment (maxfrei750) [reply scipy/stats/stats.py:426]: Ok. Would a `ValueError` be appropriate in your opinion?

Comment (mdhaber) [reply scipy/stats/stats.py:424]: Yes, I think so.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:422]: I just reconsidered this. The point that if `weights` is already a masked array, you want it to stay a masked array, and if it is not, you want to make it a regular array, right? That would be OK?
If so, how about
```
if weights isinstance(weights, np.ma.MaskedArray):
   weights = weights.astype(dtype)
else:
   weights = np.array(weights, dtype=dtype, copy=False)
```
And consider putting that in a try - except so if the user enters something wonky you can give them a helpful message.

File: scipy/stats/stats.py
Comment (maxfrei750) [reply scipy/stats/stats.py]: Resolved in 6a0d8258ddd896ebe35babc9ea583e874a139115.

File: scipy/stats/stats.py
Comment (maxfrei750) [reply scipy/stats/stats.py:422]: If I understood correctly, then that is indeed the intention. However, I just based the input conversion on that of the `a` parameter, because I thought it would be plausible to apply the same methods to both inputs.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:424]: Thanks for the `ValueError`.
On more carefuly consideration, I don't think the shapes need to be identical. I think `np.average` has the size checking you need. I would actually suggest not raising your own message because the message `np.average` provides is quite readable, e.g.
```
In [25]: np.exp(np.average(log_a, axis=0, weights=weights))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-25-612396c6a193> in <module>
----> 1 np.exp(np.average(log_a, axis=0, weights=weights))

<__array_function__ internals> in average(*args, **kwargs)

~\Anaconda3\envs\scipydev\lib\site-packages\numpy\lib\function_base.py in average(a, axis, weights, returned)
    399             if wgt.shape[0] != a.shape[axis]:
    400                 raise ValueError(
--> 401                     "Length of weights not compatible with specified axis.")
    402
    403             # setup wgt to broadcast along axis

ValueError: Length of weights not compatible with specified axis.
```

Comment (maxfrei750) [reply scipy/stats/stats.py]: Makes sense. Resolved in d01121656975514443a25aae9221b29db403ebcd.

File: scipy/stats/stats.py
Comment (maxfrei750) [reply scipy/stats/stats.py]: The suggested changes resulted in broken code, so I reverted 049d1f60634e54db506599481294961092b6918c in cd8ad1bb894343ec6d318e917bb0225b392177a2 for now.

Comment (maxfrei750) [reply scipy/stats/stats.py]: > Note that this will also need a unit test.

@rgommers and @mdhaber How many and what kinds of test would you suggest? Do all the existing tests need to be repeated with weights?

**Comment by maxfrei750** [comment]
> Note that this will also need a unit test.

@rgommers and @mdhaber How many and what kinds of test would you suggest? Do all the existing tests need to be repeated with weights?

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py]: OK. I would suggest that the four lines above express that intent more directly and on average executes with fewer `if` checks. 

There's at least two other advantage of the code above in special cases:

Suppose `weights` is input as a regular NumPy array with `dtype` `np.float64`, and `log_a` is not a masked array. If the user enters the argument `dtype=np.float64`, then:
`if not isinstance(weights, np.ndarray)` will be false
`if isinstance(a, np.ma.MaskedArray):` will be false
`weights = np.asarray(weights, dtype=dtype)` will run. This will copy `weights` unnecessarily.

Suppose `weights` is input as a regular NumPy array with `dtype` `np.float64`, and `log_a` is a masked array. If the user enters the argument `dtype=np.float64`, then:
`if not isinstance(weights, np.ndarray)` will be false
`if isinstance(a, np.ma.MaskedArray):` will be True
`weights = np.ma.asarray(weights, dtype=dtype)` will run. This converts `weights` to a masked array unnecessarily.

Not blocking, so you can leave as-is if you prefer what you have, but maybe other reviewers will weight in.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:370]: Maybe add the information about dimensions from `np.average` documentation:
> The `weights` array can either be 1-D (in which case its length must be the size of `a` along the given `axis`) or of the same shape as `a`.

Comment (maxfrei750) [reply scipy/stats/stats.py]: Good advice. Thanks! Resolved in 05833658fc4dd28e3fa2f5d5998905b6dbdaaf93.

**Comment by mdhaber** [comment]
> Do all the existing tests need to be repeated with weights?
Personally I don't think so.

> How many and what kinds of test would you suggest?

* Can you find a result of a gmean calculation somewhere online so that you can check the accuracy of your result? That would be a good test to add (even though the calculation is really simple). If you can't find a numerical result, you could check the result against the result of some other software (e.g. R).
* It would also be good to do a test where `weights` is provided as a masked array and another in which `weights` is provided as some other sort of iterable, like a list.
* Test that you get the desired error if `weights` can't be converted into an array, e.g. `weights = 'ekki'`.

File: scipy/stats/stats.py
Comment (maxfrei750) [reply scipy/stats/stats.py]: I trust your expertise and incorporated your changes in 55ff9d2.

**Comment by maxfrei750** [comment]
I added the proposed tests. However, I am not sure, whether the last test (bad input) is optimal yet. A `TypeError` makes sense. However, the error message might be cryptic for some users:
> TypeError: 1D weights expected when shapes of a and weights differ.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:372]:
```suggestion
        The weights array can either be 1-D (in which case its length must be
        the size of `a` along the given `axis`) or of the same shape as `a`.
        Default is None, which gives each value a weight of 1.0.
```
So that the reader can more easily distinguish between parameter names and English words.

**Comment by mdhaber** [comment]
> However, the error message might be cryptic for some users:

True. We could improve input validation, but I would only ask you to give a helpful message if the _new_ code fails. You could do that by starting a `try` after `if weights is not none` and raising an error if something goes wrong, but there's not much that `np.array` won't accept, and even if the user passes a `dtype` that doesn't work, `np.array` gives good error messages, so I'm not sure it's needed.

If you wanted to improve the error message for the test case you included, I would suggest adding it to `np.average` instead of this code, and that can't even be included in his PR, so I think it's out of scope : )

Normally I find myself spending more time on input validation and testing than the code itself, but I'm not sure it's required here. I think what you have checks the essential boxes, and since it's pretty simple and clean, I don't know how much more is needed.

I think the more important thing is to add a reference about how the weighted geometric mean is calculated. When that's done, and the backticks are added to the docstring, and I'll try to get another reviewer to take a look before merging.

**Comment by maxfrei750** [comment]
I think this PR should now be complete. Let me know if something else needs to be done. Thanks for your help so far.

**Comment by mdhaber** [comment]
@rlucas7 I wondered if I could attract your attention to this one. It seems straightforward and useful, so I thought it might be a nice break from larger/messier discussions!

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4483]: One might say that these should be `@parametrize`d, but I think they're already pretty short, and it's nice that each one has a meaningful name.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:420]: I just learned about np.asanyarray. i think that would do the trick here, actually. No need to change it @maxfrei750; i can.

Comment (maxfrei750) [reply scipy/stats/stats.py:420]: Feel free to aply changes as necessary.

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:424]:
```suggestion
        weights = np.asanyarray(weights, dtype=dtype)
```

File: scipy/stats/tests/test_stats.py
Comment (rlucas7) [inline scipy/stats/tests/test_stats.py:4503]: I'm confused as to why a masked array is used but no `nan`, usually these arise from a `nan_policy` argument as a result of masked particular entries that are associated w/`nan`s

File: scipy/stats/stats.py
Comment (rlucas7) [inline scipy/stats/stats.py:369]: can `weights` be negative or `0`, can they be any type or should they be an array_like of ints or positive floats?

File: scipy/stats/stats.py
Comment (rlucas7) [inline scipy/stats/stats.py:371]: if the shape of `weights` has a higher dimension than `a` what is the expected behavior?

File: scipy/stats/tests/test_stats.py
Comment (rlucas7) [reply scipy/stats/tests/test_stats.py:4483]: > One might say that these should be `@parametrize`d, but I think they're already pretty short, and it's nice that each one has a meaningful name.

I'd think an string that is unused in the test could be passed as argument w/the name/context given in the string. 

Yeah these seem ok w/o the parametrize for now.

**Review by rlucas7 — CHANGES_REQUESTED** [review]
@maxfrei750 I left a couple questions and comments please take a look.

File: scipy/stats/tests/test_stats.py
Comment (maxfrei750) [reply scipy/stats/tests/test_stats.py:4503]: If I understand you correctly, you would like to have something along the lines of:
```
a = np.array([1, 2, 3, 4, 5, np.nan])
weights = np.ma.array([2, 5, 6, 4, 3, np.nan], mask=[0, 0, 0, 0, 0, 1])
```
Is that correct?

File: scipy/stats/stats.py
Comment (maxfrei750) [reply scipy/stats/stats.py:369]: The calculation can definitively be performed with non-positive weights. Whether this actually makes sense, depends IMHO on the application of the user.

File: scipy/stats/stats.py
Comment (maxfrei750) [reply scipy/stats/stats.py:371]: The expected behavior is to mimic `numpy.average`:

If `axis` is `None` (default):
```
TypeError: Axis must be specified when shapes of a and weights differ.
```

If `axis` is not `None`:
```
TypeError: 1D weights expected when shapes of a and weights differ.
```

So basically, 

> if the shape of `weights` has a higher dimension than `a`

an error is being raised.

**Comment by maxfrei750** [comment]
@rlucas7 I tried to answer your questions. For the one concerning the test of the masked arrays, I need some more input from you. I would consider the others resolved. If you agree, then please mark the discussions as resolved.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:4503]: My interpretation of what @rlucas7 is saying is that it is more common to handle `nan`s in `scipy.stats` function than to support masked array input, and that usually the only use of masked arrays in `scipy.stats` is to handle `nan_policy='omit'`, so it is unusual that there is a test in here for masked array input when there is not a test of `nan`s in the input.

I would say that since this function doesn't have a `nan_policy`, I don't think we need a `nan` test. 

Also, since this function is not in `mstats`, I don't think this function _needs_ to work on masked arrays or have a masked array test, either. However, it doesn't hurt to show that nothing about the addition of the `weight` argument prevents this function from working properly on masked arrays.

**Review by mdhaber — APPROVED** [review]
Pretty straightforward enhancement, and LGTM!

File: scipy/stats/tests/test_stats.py
Comment (rlucas7) [reply scipy/stats/tests/test_stats.py:4503]: @maxfrei750 the description @mdhaber wrote is what I was pointing out. As @mdhaber mentioned there is nothing wrong with the function handling masked arrays-just a little unusual relative to the rest of the code base. 

>I would say that since this function doesn't have a nan_policy, I don't think we need a nan test.

I agree that we don't need a test for anything nan_policy related.

File: scipy/stats/stats.py
Comment (rlucas7) [reply scipy/stats/stats.py:369]: ok thanks for confirming. Let's leave this as is then, marking resolved.

File: scipy/stats/stats.py
Comment (rlucas7) [reply scipy/stats/stats.py:371]: ok thanks for confirming.

**Review by rlucas7 — APPROVED** [review]

**Comment by rlucas7** [comment]
@mdhaber thanks for the ping, I took another look and everything looks good so squashed and merged. 

@maxfrei750 thank you for the contribution. I'll add a note in the 1.7.0 release notes.

**Comment by mdhaber** [comment]
Thanks @maxfrei750 @rlucas7 !

**Comment by maxfrei750** [comment]
That's great news! Thanks for your help @mdhaber and @rlucas7!

**Pull request merged**
